test/common/http/http2/codec_impl_test_util.h

```diff
@@ -26,7 +26,16 @@ class TestCodecSettingsProvider {
   void onSettingsFrame(const nghttp2_settings& settings_frame) {
     for (uint32_t i = 0; i < settings_frame.niv; ++i) {
       auto result = settings_.insert(settings_frame.iv[i]);
-      ASSERT(result.second);
+      // It is possible to have duplicate settings parameters, each new parameter replaces any
+      // existing value.
+      // https://tools.ietf.org/html/rfc7540#section-6.5
+      if (!result.second) {
+        ENVOY_LOG_MISC(debug, "Duplicated settings parameter {} with value {}",
+                       settings_frame.iv[i].settings_id, settings_frame.iv[i].value);
+        settings_.erase(result.first);
```

Comment:
It seems preferable to use a map instead now that uniqueness checks are not required.

Reply:
It depends on what the intent of this helper was.  If it was to detect erroneous accidental double settings in tests, then the original was best.

It's a bit strange that we do a set of nghttp2_settings_entry instead of a map<settings_id, value> and fail when setting_id is duplicated.  I can see us doing the nghttp2_settings_entry if we're trying to avoid errorneously sending the exact same setting_id, value pair multiple times.

Reply:
The intent was to simply provide a helper to fetch SETTINGS parameters received from a peer. The use of the set was based on my misunderstanding of the RFC; the goal was to be very explicit about the uniqueness requirement through the type used.

Reply:
Following up on Antonio's point about accidentally sending duplicate pairs, we _should_ add a test that validates that condition does not happen, but this utility should be generic enough to support both the existing unit tests, fuzz tests and to-be-written duplicate parameter tests.

Reply:
Changed to an unordered-map. this isn't relevant to `sendSettings`, which fwiw parses configuration settings for your codecs. (And we can impose restrictions like do not duplicate settings in your config). I can write a test that sends a settingsframe containing duplicate settings in a frame (this would have always worked against a real envoy, since this is just test utility)